src/functions/routers/mod.ts

```diff
@@ -0,0 +1,125 @@
+import type { ActionHandler } from "./types.ts";
+
+/**
+ * Define an actions "router" and its input and output parameters for use in a Slack application. The ActionsRouter will route incoming action events to action-specific handlers.
+ * @param {SlackFunction<InputParameters, OutputParameters, RequiredInput, RequiredOutput>} func Reference to your previously-created SlackFunction, defined via DefineFunction
+ * @returns {ActionsRouter}
+ */
+export const BlockActionsRouter = <
+  InputParameters extends ParameterSetDefinition,
+  OutputParameters extends ParameterSetDefinition,
+  RequiredInput extends PossibleParameterKeys<InputParameters>,
+  RequiredOutput extends PossibleParameterKeys<OutputParameters>,
+>(
+  func: SlackFunction<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  >,
+) => {
+  return new ActionsRouter(func);
+};
+
+class ActionsRouter<
+  InputParameters extends ParameterSetDefinition,
+  OutputParameters extends ParameterSetDefinition,
+  RequiredInput extends PossibleParameterKeys<InputParameters>,
+  RequiredOutput extends PossibleParameterKeys<OutputParameters>,
+> {
+  private func: SlackFunction<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  >;
+  private routes: Array<
+    Array<string[] | RegExp | ActionHandler<typeof this.func.definition>>
+  >;
+
+  constructor(
+    func: SlackFunction<
+      InputParameters,
+      OutputParameters,
+      RequiredInput,
+      RequiredOutput
+    >,
+  ) {
+    this.func = func;
+    this.routes = [];
+  }
+
+  /**
+   * Add an action handler for something that can matche an action_id.
+   */
+  add(
+    action_id: string | string[] | RegExp,
+    handler: ActionHandler<typeof this.func.definition>,
+  ): ActionsRouter<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  > {
+    if (typeof action_id === "string") {
+      action_id = [action_id];
+    }
+    this.routes.push([action_id, handler]);
+    return this;
+  }
+
+  /**
+   * Returns a method handling routing of action payloads to the appropriate action handler.
+   * The output of export() should be attached to the `blockActions` export of your function.
+   */
+  export(): ActionHandler<typeof this.func.definition> {
+    return async (context) => {
+      const action = context.action.action_id as string;
+      const handler = this.matchHandler(action);
+      if (handler === null) {
+        // TODO: what do in this case?
+        // perhaps the user typo'ed the action id when registering their handler or defining their block.
+        console.warn(
+          `Received block action payload with action_id=${action} but this app has no action handler defined to handle it!`,
+        );
+        return;
+      }
+      if (handler.constructor.name === "AsyncFunction") {
+        return await handler(context);
+      } else {
+        return handler(context);
+      }
```

Comment:
For my own education — is it a problem to await a sync function here? I understand it'll treat the sync function as a promise and push the sync fn execution to a microtask queue, but will that have real impact? 🤔

Reply:
Probably not, other than linter warnings 🤷 will make this change.